### ISSUE TITLE
Module not compatible with Odoo V9

### ISSUE THREAD

**Issue by looppi** [avoinsystems/l10n_fi_invoice#2]
1. Installing module fails with XML errors. My [commit](https://github.com/looppi/l10n_fi_invoice/commit/be9f2840a3959598b7a99fb71d75ca237f907944) solved the installation problem, but although the module installs ok, it doesn't really work.
2. After installing, any action with invoices results and error. Seems to me that the data model has changed and therefore the code fails. With a quick look there are only few cases where bank_ids are referenced in V9 codebase.

```
Odoo Server Error

Traceback (most recent call last):
  File "/opt/odoo/openerp/http.py", line 599, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/opt/odoo/openerp/http.py", line 636, in dispatch
    result = self._call_function(**self.params)
  File "/opt/odoo/openerp/http.py", line 315, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/opt/odoo/openerp/service/model.py", line 113, in wrapper
    return f(dbname, *args, **kwargs)
  File "/opt/odoo/openerp/http.py", line 308, in checked_call
    result = self.endpoint(*a, **kw)
  File "/opt/odoo/openerp/http.py", line 887, in __call__
    return self.method(*args, **kw)
  File "/opt/odoo/openerp/http.py", line 465, in response_wrap
    response = f(*args, **kw)
  File "/opt/odoo/addons/web/controllers/main.py", line 971, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "/opt/odoo/addons/web/controllers/main.py", line 963, in _call_kw
    return checked_call(request.db, *args, **kwargs)
  File "/opt/odoo/openerp/service/model.py", line 113, in wrapper
    return f(dbname, *args, **kwargs)
  File "/opt/odoo/addons/web/controllers/main.py", line 962, in checked_call
    return getattr(request.registry.get(model), method)(request.cr
, request.uid, *args, **kwargs)
  File "/opt/odoo/openerp/api.py", line 223, in wrapper
    return old_api(self, *args, **kwargs)
  File "/opt/odoo/openerp/api.py", line 318, in old_api
    result = method(recs, *args, **kwargs)
  File "/opt/odoo/addons/account/models/account_payment.py", line 242, in default_get
    invoice_defaults = self.resolve_2many_commands('invoice_ids', rec.get('invoice_ids'))
  File "/opt/odoo/openerp/api.py", line 221, in wrapper
    return new_api(self, *args, **kwargs)
  File "/opt/odoo/openerp/api.py", line 454, in new_api
    result = method(self._model, cr, uid, *args, **kwargs)
  File "/opt/odoo/openerp/models.py", line 5191, in resolve_2many_commands
    for record in other_model.read(cr, uid, record_ids, fields=fields, context=context):
  File "/opt/odoo/openerp/api.py", line 223, in wrapper
    return old_api(self, *args, **kwargs)
  File "/opt/odoo/openerp/models.py", line 3168, in read
    result = BaseModel.read(records, fields, load=load)
  File "/opt/odoo/openerp/api.py", line 221, in wrapper
    return new_api(self, *args, **kwargs)
  File "/opt/odoo/openerp/models.py", line 3214, in read
    values[name] = field.convert_to_read(record[name], use_name_get)
  File "/opt/odoo/openerp/models.py", line 5668, in __getitem__
    return self._fields[key].__get__(self, type(self))
  File "/opt/odoo/openerp/fields.py", line 805, in __get__
    self.determine_value(record)
  File "/opt/odoo/openerp/fields.py", line 912, in determine_value
    self.compute_value(recs)
  File "/opt/odoo/openerp/fields.py", line 867, in compute_value
    self._compute_value(records)
  File "/opt/odoo/openerp/fields.py", line 857, in _compute_value
    getattr(records, self.compute)()
  File "/opt/odoo/openerp/api.py", line 221, in wrapper
    return new_api(self, *args, **kwargs)
  File "/opt/odoo/openerp/api.py", line 389, in new_api
    result = [method(rec, *args, **kwargs) for rec in self]
  File "/opt/external-addons/l10n_fi_invoice/model/account_invoice.py", line 50, in _compute_barcode_string
    self.company_id.bank_ids and self.company_id.bank_ids[0]
AttributeError: 'res.company' object has no attribute 'bank_ids'
```

**Comment by mlaitinen** [comment]
1. Odoo 9 doesn't support referring to element by its `string` attribute, so that is why the install fails. Your fixed xpath expressions look fine to me.
2. A quick glimpse reveals that `res.company` doesn't indeed have the field `bank_ids` any more. It was probably redundant because `res.partner` already has that field.
   Without actually testing this, I believe the problem might be solved by changing `self.company_id.bank_ids` to `self.company_id.partner_id.bank_ids`.

**Comment by looppi** [comment]
It seems that the problem number 2 doesn't go away just by changing couple of lines. My impression after wrestling a while with the report code is that somewhat the whole data model has changed or at least everything has been renamed.

For example there are numerous references to _o_ variable, which should have company info and so on, with my tests all I got was NoneType for this _o_.

Stack trace coming from trying to include company logo to the invoice:

```
  File "/opt/odoo/openerp/addons/base/ir/ir_qweb.py", line 191, in eval
    return qwebcontext.safe_eval(expr)
  File "/opt/odoo/openerp/addons/base/ir/ir_qweb.py", line 104, in safe_eval
    return eval(expr, None, locals_dict, nocopy=True, locals_builtins=True)
  File "/opt/odoo/openerp/tools/safe_eval.py", line 298, in safe_eval
    return eval(c, globals_dict, locals_dict)
  File "", line 1, in <module>
QWebException: "'NoneType' object has no attribute 'company_id'" while evaluating
"'data:image/png;base64,%s' % o.company_id.logo"
```

**Comment by mlaitinen** [comment]
Ok, I got this sorted out. In Odoo v8 you call `translate_doc` to translate QWeb reports. `translate_doc` sets the document in a variable called `o`. In Odoo v9 you give the language in the `t-lang` parameter instead of `translate_doc` and in that same loop the document is set in the `doc` variable.

See: https://github.com/avoinsystems/l10n_fi_invoice/compare/8.0...looppi:9.0#diff-3053d9749c0120f41757514bda97208aR500 and (example:) https://github.com/odoo/odoo/blob/9.0/addons/sale/views/report_saleorder.xml.

So the solution is to just rename all `o.` to `doc.` in `l10n_fi_invoice.report_invoice_finnish_document`.

**Comment by looppi** [comment]
Thanks a lot! Now I got it to actually print the PDF without errors. Some minor tweaks are still required, but the basic functionality is now working.

How do we continue this? Do I create a pull request containing the changes?

**Comment by mlaitinen** [comment]
Pull request sounds good. I just created a niner branch for [avoinsystems/l10n_fi_invoice](https://github.com/avoinsystems/l10n_fi_invoice/commits/9.0).

**Comment by mlaitinen** [comment]
Closed by #3
